sklearn/metrics/cluster/_supervised.py

```diff
@@ -138,6 +145,209 @@ def contingency_matrix(labels_true, labels_pred, *, eps=None, sparse=False):
+      >>> pair_confusion_matrix([0, 0, 1, 1], [0, 0, 1, 1])
+      [[8, 0], [0, 4]]
+      >>> pair_confusion_matrix([0, 0, 1, 1], [1, 1, 0, 0])
+      [[8, 0], [0, 4]]
+
+    Labelings that assign all classes members to the same clusters
+    are complete but may be not always pure, hence penalized, and
+    have some off-diagonal non-zero entries:
+
+      >>> pair_confusion_matrix([0, 0, 1, 2], [0, 0, 1, 1])  # doctest: +ELLIPSIS
+      [[8, 2], [0, 2]]
+
+    The matrix is not symmetric:
+
+      >>> pair_confusion_matrix([0, 0, 1, 1], [0, 0, 1, 2])  # doctest: +ELLIPSIS
+      [[8, 0], [2, 2]]
+
+    If classes members are completely split across different clusters, the
+    assignment is totally incomplete, hence the matrix has all zero
+    diagonal entries:
+
+      >>> pair_confusion_matrix([0, 0, 0, 0], [0, 1, 2, 3])
+      [[0, 0], [12, 0]]
+
+    References
+    ----------
+
+    .. [Hubert1985] L. Hubert and P. Arabie, Comparing Partitions,
+      Journal of Classification 1985
+      https://link.springer.com/article/10.1007%2FBF01908075
+
+    See also
+    --------
+    adjusted_pair_confusion_matrix: Adjusted Rand Score
+    adjusted_mutual_info_score: Adjusted Mutual Information
+
+    """
+    labels_true, labels_pred = check_clusterings(labels_true, labels_pred)
+    n_samples = np.int64(labels_true.shape[0])
+    n_classes = np.unique(labels_true).shape[0]
+    n_clusters = np.unique(labels_pred).shape[0]
+
+    # Special limit cases: no clustering since the data is not split;
+    # or trivial clustering where each document is assigned a unique cluster.
+    # These are perfect matches hence all pairs are counted either under
+    # D00 or D11.
+    if n_classes == n_clusters == 1:
+        return [[0,0], [0,n_samples * (n_samples - 1)]]
+    elif n_classes == n_clusters == 0:
+        return [[0,0], [0,0]]
+    elif n_classes == n_clusters == n_samples:
+        return [[n_samples * (n_samples - 1),0], [0,0]]
+
+    # Computation using the contingency data
+    contingency = contingency_matrix(labels_true, labels_pred, sparse=True, dtype=np.int64)
+    n_c = np.ravel(contingency.sum(axis=1))
+    n_k = np.ravel(contingency.sum(axis=0))
+    #rows,cols = contingency.nonzero()
+    #D11 = sum(n_ij*(n_ij-1) for n_ij in contingency.data)
+    Sum_Squares = sum(n_ij*n_ij for n_ij in contingency.data)
+    D11 = Sum_Squares - n_samples
+    #D01 = sum(contingency[c,k] * (n_k[k] - contingency[c,k]) for c,k in zip(rows,cols))
+    D01 = sum(contingency.dot(n_k)) - Sum_Squares
+    #D10 = sum(contingency[c,k] * (n_c[c] - contingency[c,k]) for c,k in zip(rows,cols))
+    D10 = sum(contingency.transpose().dot(n_c)) - Sum_Squares
+    #D00 = sum(contingency[c,k] * (n_samples - n_c[c] - n_k[k] + contingency[c,k]) for c,k in zip(rows,cols))
+    D00 = n_samples * n_samples - D01 - D10 - Sum_Squares
+    return [[D00, D01], [D10, D11]]
+
+
+def rand_score(labels_true, labels_pred):
```

Comment:
this function needs tests in `sklearn/metrics/cluster/tests/test_supervised.py`, and inclusion in `sklearn/metrics/cluster/tests/test_common.py`